prog/vm/github_runner.rb

```diff
@@ -62,6 +62,31 @@ def pick_vm
     vm_st.subject
   end
 
+  def find_closest_docker_mirror_server
+    compatible_servers = DockerRegistryMirrorServer.all.select { |server| server.vm.arch == vm.arch }
+    return nil if compatible_servers.empty?
+
+    same_vm_host_server = nil
+    same_data_center_server = nil
+    same_region_server = nil
+    fallback_server = nil
+
+    compatible_servers.each do |server|
+      if server.vm.vm_host.id == vm.vm_host_id
+        same_vm_host_server = server
+        break
+      elsif server.vm.vm_host.data_center == vm.vm_host.data_center
+        same_data_center_server = server
+      elsif server.vm.vm_host.data_center&.split("-")&.first == vm.vm_host.data_center&.split("-")&.first
+        same_region_server = server
+      else
+        fallback_server = server
+      end
+    end
+
+    same_vm_host_server || same_data_center_server || same_region_server || fallback_server
```

Comment:
It's up to you. You can use a scoring system similar to our allocator. One advantage is that it picks randomly if the servers have the same criteria.

```suggestion
     compatible_servers.map do |server|
      if server.vm.vm_host.id == vm.vm_host_id
        [0, server]
      elsif server.vm.vm_host.data_center == vm.vm_host.data_center
        [1, server]
      elsif server.vm.vm_host.data_center&.split("-")&.first == vm.vm_host.data_center&.split("-")&.first
        [2, server]
      else
        [3, server]
      end
    end.min_by { _1.first + rand }.last
```